src/core/expr/fexpr_fillna.cc

```diff
@@ -0,0 +1,132 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include "documentation.h"
+#include "expr/fexpr_func.h"
+#include "expr/eval_context.h"
+#include "python/xargs.h"
+#include "parallel/api.h"
+namespace dt {
+namespace expr {
+
+
+class FExpr_FillNA : public FExpr_Func {
+  private:
+    ptrExpr arg_;
+    bool reverse_;
+    size_t : 56;
+
+  public:
+    FExpr_FillNA(ptrExpr &&arg, bool reverse)
+        : arg_(std::move(arg)),
+          reverse_(reverse)
+         {}
+
+
+    std::string repr() const override {
+      std::string out = "fillna";
+      out += '(';
+      out += arg_->repr();
+      out += ", reverse=";
+      out += reverse_? "True" : "False";
+      out += ')';
+      return out;
+    }
+
+
+    template <bool REVERSE>
+    static RowIndex fill_rowindex(Column& col, const Groupby& gby) {
+      Buffer buf = Buffer::mem(static_cast<size_t>(col.nrows()) * sizeof(int32_t));
+      auto indices = static_cast<int32_t*>(buf.xptr());
+
+      dt::parallel_for_dynamic(
+        gby.size(),
+        [&](size_t gi) {
+          size_t i1, i2;
+          gby.get_group(gi, &i1, &i2);
+          size_t fill_id = REVERSE? i2 - 1 : i1;
+
+          if (REVERSE) {
+            for (size_t i = i2; i-- > i1;) {
+              size_t is_valid = col.get_element_isvalid(i);
+              fill_id = is_valid? i : fill_id;
+              indices[i] = static_cast<int32_t>(fill_id);
+            }
+          } else {
+            for (size_t i = i1; i < i2; ++i) {
+              size_t is_valid = col.get_element_isvalid(i);
+              fill_id = is_valid? i : fill_id;
+              indices[i] = static_cast<int32_t>(fill_id);
+            }
+          }
+
+        }
+      );
+
+      return RowIndex(std::move(buf), RowIndex::ARR32|RowIndex::SORTED);
+    }
+
+
+    Workframe evaluate_n(EvalContext &ctx) const override {
+      Workframe wf = arg_->evaluate_n(ctx);
+      Groupby gby = Groupby::single_group(wf.nrows());
+      if (ctx.has_groupby()) {
+        wf.increase_grouping_mode(Grouping::GtoALL);
+        gby = ctx.get_groupby();
+      }
+
+      for (size_t i = 0; i < wf.ncols(); ++i) {
+        Column coli = wf.retrieve_column(i);
+        bool is_grouped = ctx.has_group_column(
+                            wf.get_frame_id(i),
+                            wf.get_column_id(i)
+                          );
+        if (!is_grouped){
```

Comment:
@oleksiyskononenko is the null count check irrelevant? I thought it would skip if there were no nulls? or does the `is_grouped` boolean somehow cover that?

Reply:
It would be nice to have this check if it comes for free. However, to actually calculate the number of nulls in the column one needs to loop over the column data in a loop that is kind of similar to the loop which we implement in `fillna()`. Since most of the columns have some missing data, it means that in most of the cases we will have to go through the data twice. My feeling that in this case it is better to remove the check.

Reply:
One thing we can actually do here is to check if the stats is already computed and the number of NA's is already known, something similar to https://github.com/h2oai/datatable/blob/main/src/core/column/column_impl.cc#L175

Reply:
I assumed it was already precomputed and available?

Reply:
Nope, until touched it is not there and needs to be computed. Let me push an update, so that we reuse it if available.